paddle/fluid/imperative/layer.cc

```diff
@@ -240,189 +350,452 @@ std::unique_ptr<VarBase> VarBase::NewVarBase(const platform::Place& dst_place,
+  // Build backward ins and outs
+  bwd_ins_.resize(grad_op_num);
+  bwd_outs_.resize(grad_op_num);
 
-    for (const auto& it : grad_input_vars_[k]) {
-      auto& grad_invars = grad_invars_map[it.first];
-      grad_invars.reserve(it.second.size());
-      for (const VarBase* grad_inp : it.second) {
-        PADDLE_ENFORCE_NOT_NULL(grad_inp->var_, "op %s input %s nullptr",
-                                grad_op_desc->Type(), grad_inp->Name());
+  for (size_t i = 0; i < grad_op_num; ++i) {
+    for (auto& grad_ins : grad_op_descs_[i]->Inputs()) {
+      if (grad_ins.second.empty()) continue;
+      auto& bwd_in = bwd_ins_[i][grad_ins.first];
+      bwd_in.reserve(grad_ins.second.size());
 
-        grad_invars.emplace_back(grad_inp->var_.get());
+      for (auto& grad_in_var_name : grad_ins.second) {
+        auto iter = grad_to_var.find(grad_in_var_name);
+
+        if (iter != grad_to_var.end()) {
+          // If it is a grad var, find its coresponding forward var
+          auto& fwd_var_name = iter->second;
+          auto fwd_var_iter = name_to_var.find(fwd_var_name);
+          PADDLE_ENFORCE(fwd_var_iter != name_to_var.end(),
+                         "Cannot find forward variable named %s", fwd_var_name);
+
+          bwd_in.emplace_back((*(fwd_var_iter->second))->GradVarBase());
+        } else {
+          // If it is a forward var, just add it
+          auto fwd_var_iter = name_to_var.find(grad_in_var_name);
+          PADDLE_ENFORCE(fwd_var_iter != name_to_var.end(),
+                         "Cannot find forward variable named %s",
+                         grad_in_var_name);
+          bwd_in.emplace_back(*(fwd_var_iter->second));
+        }
+
+        VLOG(2) << "Set backward input " << grad_ins.first << " of " << type_
+                << " to be " << bwd_in.back()->Name();
       }
     }
 
-    for (const auto& it : tmp_grad_outputs[k]) {
-      auto& grad_outvars = grad_outvars_map[it.first];
-      grad_outvars.reserve(it.second.size());
-      for (VarBase* grad_out : it.second) {
-        PADDLE_ENFORCE_NOT_NULL(grad_out->var_, "op %s output %s nullptr",
-                                grad_op_desc->Type(), grad_out->Name());
+    for (auto& grad_outs : grad_op_descs_[i]->Outputs()) {
+      if (grad_outs.second.empty()) continue;
+      auto& bwd_out = bwd_outs_[i][grad_outs.first];
+      bwd_out.reserve(grad_outs.second.size());
+
+      for (auto& grad_out_var_name : grad_outs.second) {
+        auto iter = grad_to_var.find(grad_out_var_name);
+        PADDLE_ENFORCE(iter != grad_to_var.end(),
+                       "Cannot find output of input grad %s in op %s",
+                       grad_out_var_name, type_);
+        auto fwd_var_iter = name_to_var.find(iter->second);
+        PADDLE_ENFORCE(fwd_var_iter != name_to_var.end(),
+                       "Cannot find forward variable named %s", iter->second);
+        bwd_out.emplace_back((*(fwd_var_iter->second))->GradVarBase());
```

Comment:
Not needed. Maybe it is confused to set `shape` in `VarBase` because `VarBase` is not always a `LoDTensor`. It may be a `SelectedRows` in the future, and `SelectedRows` does not support `shape`.